imports.py

```diff
@@ -12,193 +10,283 @@
-		else:
-			print('Invalid code')
-
-	def lookup_code(self, system: URI, code: str):
-		"""Lookup code within code system"""
-		response: Dict = self.do_hapi_request(f'CodeSystem/$lookup?system={system}&code={code}')
-		decode_lookup_response(response)
+class CacheFhir:
+    """ missing an invalidation mechanism, use with caution """
+    
+    @staticmethod
+    def encode_filename_str(x):
+        """ Convert such that can create friendly system filename.
+            The issue here is backwards from what you might expect.
+            
+        """
+        #delim = '*'
+        #return x.replace('/', f'{delim}slash')
+        return x.replace('/', '*slash')
+
+    @staticmethod
+    def decode_filename_str(x):
+        """Convert from friendly system filename"""
+        #return x.replace(f'{delim}slash', '/')
+        return x.replace('*slash', '/')
+
+    def url_and_params_to_filename(url: str, params: Dict) -> str:
+        """Convert to filename"""
+        x = url
+        if params:
+            x += '___params_'
+            for k, v in params.items():
+                x += '__' + k + '_' + v
+        filename = CacheFhir.encode_filename_str(x) + '.json'
+        return filename
+    
+    def cache_write(url: str, params: Dict, response: Dict):
+        """Write to cache"""
+        filename = CacheFhir.url_and_params_to_filename(url, params)
+        if not os.path.exists(CACHE_DIR):
+            os.makedirs(CACHE_DIR)
+        path = os.path.join(CACHE_DIR, filename)
+        with open(path, 'w') as f:
+            json.dump(response, f, indent=4)
+
+    def cache_read(url: str, params: Dict) -> Union[Dict, None]:
+        """Read from cache"""
+        filename = CacheFhir.url_and_params_to_filename(url, params)
+        path = os.path.join(CACHE_DIR, filename)
+        if os.path.exists(path):
+            with open(path, 'r') as f:
+                data: Dict = json.load(f)
+                return data
+        return None
+
 
 
+
+class TimsClient:
+    """TIMS Client"""
+
+    def __init__(self, base_url: str, use_cache=False, verbose=False):
+        self.base_url = base_url
+        self.use_cache = use_cache
+        self.verbose=verbose
+        self.http_client = httpx.Client()
+        
+    def is_response_ok(self, response: Response) -> bool:
+        """Decode status"""
+        if response is None:
+            print("NO RESPONSE")
+            return False
+        else:
+            if  200 <= response.status_code < 300:
+                if self.verbose:
+                    print(f"HTTP STATUS OK {response.status_code}")
+                return True
+            else:
+                print(f"error: {response.status_code}")
+                return False
+
+    def do_hapi_get_request(self, internal_path: str, in_params: Dict = None) -> Dict:
+        """GET request"""
+        url = self.base_url + internal_path
+        response_dict = None
+        if self.use_cache:
+            print(" * USING CACHE * ")
```

Comment:
I didn't want to see good results on accident, so I made it obvious when we're caching. The thinking is we should run from the server for real when saving the output.

Reply:
Fine with this for now. My thought is just to have as little on the screen as possible to demonstrate the functionality.  Of course I don't want to rely on the cache simply as a workaround for long or buggy server responses; that would be just wrong. Mostly I have it here for debugging purposes anyway, so I suppose that for future 'release' PDFs (or when we save the notebook itself), it'll be w/ caching off.